README.md
<a href="http://badge.fury.io/rb/stripe-cli"><img style="float:right;" src="https://badge.fury.io/rb/stripe-cli.png" alt="Gem Version" height="18"></a>
<a href="https://codeclimate.com/repos/52773e8fc7f3a3121a004455/feed"><img style="float:right;margin-right:10px;" src="https://codeclimate.com/repos/52773e8fc7f3a3121a004455/badges/67e5d04281479d6f8cb3/gpa.png"></a>
# Stripe::CLI
stripe-cli is the command line interface to [Stripe](https://stripe.com).

Offering a friendly, intuitive interface to the entire API via a git-style command suite.

With an emphasis on convenience and productivity and a commitment to pretty, colorful, but most of all READABLE output.


Uses [AwesomePrint](https://github.com/michaeldv/awesome_print) to provide very readable output on your command line.

![example output](./output.png)

## Installation

    $ gem install stripe-cli

## Configuration

For authentication, pass your secret key using the `-k` or `--key` option

    $ stripe events list -k sk_test_abcdef123456

To use a specific api version, pass in the `-v` or `--version` option

    $ stripe balance_transactions list -v "2014-08-04"

Explicitly passing in configuration every time is tedious. Avoid it by creating a configuration file that contains this data.

Stripe-CLI will check your home folder (**~/**) for the existance of a `.stripecli` file that conforms to the following example

![example config file](./example.png)

If this file exists, the specified default configuration is picked up and used automatically.

The current directory is now also checked for the existance of a `.stripecli` file. In which case, that file is used instead.

As shown in the above example, you may set multiple configuration **environments** in a single file. Just make sure to specify a default.

The default **environment** may be overridden by passing in the `-e` or `--env` option to any command

    $ stripe customers find cust_123 --env=live

If you manage separate Stripe accounts for several projects, I suggest creating a simple `.stripecli` file
at the root of each project declaring the test api key for the appropriate Stripe account.

For example `./.stripecli`:

    key = sk_test_abcdef123456

or, if you'd rather:

    default = test
    [test]
      key = sk_test_abcdef123456
    [live]
      key = sk_live_ghijkl789101

If you choose to go this route, make sure to add `.stripecli` to your `.gitignore` files.

### Date/Time

By default, Date/Time stamps are converted automatically to your local machine time. No more Epoch/Unix timestamp garbage!

If you'd prefer to have time Date/Time stamps converted to UTC time, set the config option `dates` or `-d` to `utc`. You can also set it to `unix` to see the original Unix timestamp passed from the API. By default, it uses `local`.

### Dollars vs. Cents

By default stripe-cli expects currency amounts to be specified in **dollars**. This behavior maybe unexpected to veteran Stripe users since the raw API expects **cents**.

To change the expected currency units from **dollars** to **cents**, set the config option `dollar_amounts` to `false` in your `.stripecli` file

    dollar_amounts = false

or

    [test]
      dollar_amounts = false

If not set, this option defaults to **true** meaning that `amounts` should be in **dollars**

If you need to adjust this setting temporarily, there are global flags which maybe passed to any command

    $ stripe charge create --no-dollar-amounts
    Amount In Cents: __

    $ stripe charge create --dollar-amounts
    Amount In Dollars: __



## General Features & Flexibility

Any parameters accepted by the [stripe api](https://stripe.com/docs/api) are acceptable options to pass into commands, including metadata and the `expand` parameter.

    $ stripe charges create  --amount=9.98  --description=a\ must\ have  --number=4242424242424242  --cvc=123  --exp-month=10  --exp-year=2020

or

    $ stripe charges create  --amount=16.72  --token=tok_abc123  --metadata=foo:bar meta:baz --expand=balance_transaction

or

    $ stripe charges create  --amount=1  --customer=cus_def456  --no-capture

### Interactive Menus

Passing NO (or partial) arguments to any operation, will trigger an interactive menu

    $ stripe charge create
    Amount In Dollars: __

or

    $ stripe charge create  --amount=99
    Name on Card: __

or

    $ stripe token create card --card=number:4242424242424242 name:Uther\ Pendragon cvc:911 exp-month:02
    Expiration Year: __

### cursor-based pagination

for all `list` operations

    Options:
      [--starting-after=OBJECT_ID]  # The ID of the last object in the previous paged result set.
      [--ending-before=OBJECT_ID]   # The ID of the first object in the previous paged result set, when paging backwards through the list.
      [--limit=LIMIT]               # a limit on the number of resources returned, between 1 and 100


Though Stripe advocates using cursor-based pagination, offset-based pagination is still supported.

e.g. fetching a second page

    stripe events list --count=10 --offset=10

### Exception Recovery

Api errors are rescued and their messages displayed for you to read.  No more `barfing` to `stdout`

![error rescue example](./error_message_display.png)

## Usage


    Commands:
      stripe balance         # show currently available & pending balance amounts
      stripe cards           # find, list, create, & delete cards for both customers & recipients
      stripe charges         # find, list, create, capture, & refund charges
      stripe coupons         # find, list, create, & delete coupons
      stripe customers       # find, list, create, & delete customers
      stripe events          # find & list events
      stripe help [COMMAND]  # Describe available commands or one specific command
      stripe invoice_items   # find, list, create, & delete invoice items
      stripe invoices        # find, list, pay, & close invoices
      stripe plans           # find, list, create, & delete plans
      stripe recipients      # find, list, create, & delete recipients
      stripe refunds         # find, list, & create refunds
      stripe subscriptions   # find, list, create, cancel & reactivate multiple subscriptions per customer
      stripe tokens          # find & create tokens for bank accounts & credit cards
      stripe transactions    # find & list balance transactions
      stripe transfers       # find, list, & create transfers


    Global Options:
      [-k | --key=KEY]         # One of your API secret keys, provided by Stripe
      [-e | --env=ENV]         # expects a `~/.stripecli` file with section headers for any string passed into it
      [-v | --version=VERSION] # Stripe API version-date. Looks like `YYYY-MM-DD`
      [-d | --dates=DATES]     # Date Style. It should be either local, utc, or unix. Defaults to local.
      [--dollar-amounts]       # configuration flag setting expected currency units to dollars
      [--no-dollar-amounts]    # configuration flag setting expected currency units to cents
      [--strip-nils]           # use this flag to strip nil-valued attributes from the output
      [--expand=one two three] # one or more ID properties of the response you'd like expanded into full objects

## Command Documentation

- [Charges](#charges)
    - [Capture](#charge-capture)
    - [Create](#charge-create)
    - [Find](#charge-find)
    - [List](#charges-list)
    - [Refund](#charge-refund)
- [Tokens](#tokens)
    - [Create](#token-create)
    - [Find](#token-find)
- [Cards](#cards)
    - [Create](#card-create)
    - [Delete](#card-delete)
    - [Find](#card-find)
    - [List](#cards-list)
- [Customers](#customers)
    - [Create](#customer-create)
    - [Delete](#customer-delete)
    - [Find](#customer-find)
    - [List](#customers-list)
- [Subscriptions](#subscriptions)
    - [Cancel](#subscription-cancel)
    - [Create](#subscription-create)
    - [Find](#subscription-find)
    - [List](#subscriptions-list)
    - [Reactivate](#subscription-reactivate)
- [Invoices](#invoices)
    - [Close](#invoice-close)
    - [Find](#invoice-find)
    - [List](#invoices-list)
    - [Pay](#invoice-pay)
    - [Upcoming](#invoice-upcoming)
- [Invoice Items](#invoice-items)
    - [Find](#invoice-items-find)
    - [List](#invoice-items-list)
    - [Create](#invoice-items-create)
    - [Delete](#invoice-items-delete)
- [Plans](#plans)
    - [Create](#plan-create)
    - [Delete](#plan-delete)
    - [Find](#plan-find)
    - [List](#plans-list)
- [Coupons](#coupons)
    - [Create](#coupon-create)
    - [Delete](#coupon-delete)
    - [Find](#coupon-find)
    - [List](#coupons-list)
- [Events](#events)
    - [Find](#event-find)
    - [List](#events-list)
- [BalanceTransactions](#balancetransactions)
    - [Find](#transaction-find)
    - [List](#transactions-list)
- [Recipients](#recipients)
    - [Create](#recipient-create)
    - [Delete](#recipient-delete)
    - [Find](#recipient-find)
    - [List](#recipients-list)
- [Refunds](#refunds)
    - [Create](#refund-create)
    - [Find](#refund-find)
    - [List](#refunds-list)
- [Transfers](#transfers)
    - [Create](#transfer-create)
    - [Find](#transfer-find)
    - [List](#transfers-list)

<!-- end toc -->


### Charges

    Commands:
      stripe charges capture ID      # Capture a charge
      stripe charges create          # Create a charge
      stripe charges find ID         # Find a charge
      stripe charges help [COMMAND]  # Describe subcommands or one specific subcommand
      stripe charges list            # List charges (optionally by customer_id)
      stripe charges refund ID       # Refund a charge

#### Charge Capture

    Usage:
      stripe charges capture ID

    Capture a charge

#### Charge Create

    Usage:
      stripe charges create

    Options:
          [--customer=CUSTOMER]                           # The ID of an existing customer to charge
          [--token | --card=CARD]                         # credit card Token or ID. May also be created interactively.
          [--number | --card-number=CARD_NUMBER]          # credit card number. usually 16 digits long
          [--exp-month | --card-exp-month=CARD_EXP_MONTH] # Two digit expiration month of card
          [--exp-year | --card-exp-year=CARD_EXP_YEAR]    # Four digit expiration year of card
          [--cvc | --card-cvc=CARD_CVC]                   # Three or four digit security code located on the back of card
          [--name | --card-name=CARD_NAME]                # Cardholder's full name as displayed on card
          [--amount=N]                                    # Charge amount in dollars (or cents when --no-dollar-amounts)
          [--currency=CURRENCY]                           # 3-letter ISO code for currency
                                                          # Default: usd
          [--description=DESCRIPTION]                     # Arbitrary description of charge
          [--capture], [--no-capture]                     # Whether or not to immediately capture the charge. Uncaptured charges expire in 7 days
                                                          # Default: true
          [--statement-description=STATEMENT_DESCRIPTION] # Displayed alongside your company name on your customer's card statement (15 character max)
          [--receipt-email=RECEIPT_EMAIL]                 # Email address to send receipt to. Overrides default email settings.
          [--metadata=key:value]                          # A key/value store of additional user-defined data

    Create a charge

#### Charge Find

    Usage:
      stripe charges find ID

    Find a charge

#### Charges List

    Usage:
      stripe charges list

    Options:
          [--starting-after=STARTING_AFTER]  # The ID of the last object in the previous paged result set. For cursor-based pagination.
          [--ending-before=ENDING_BEFORE]    # The ID of the first object in the previous paged result set, when paging backwards through the list.
          [--limit=LIMIT]                    # a limit on the number of resources returned, between 1 and 100
          [--offset=OFFSET]                  # the starting index to be used, relative to the entire list
          [--count=COUNT]                    # deprecated: use limit
          [--customer=CUSTOMER]              # ID of customer who's charges we want to list

    List charges (optionally by customer_id)

#### Charge Refund

    Usage:
      stripe charges refund ID

    Options:
          [--amount=N]                    # Refund amount in dollars. (or cents when --no-dollar-amounts) defaults to entire charged amount
          [--refund-application-fee],     # or,
          [--no-refund-application-fee]   # Whether or not to refund the application fee
          [--metadata=key:value]          # a key/value store of additional user-defined data

    Refund a charge

### Tokens

    Commands:
      stripe tokens create TYPE           # Create a new token of type TYPE(card or account)
      stripe tokens find ID               # Find a Token
      stripe tokens help [COMMAND]        # Describe subcommands or one specific subcommand

#### Token Create

    Usage:
      stripe token create TYPE

    Options:
          [--card=key:value]                              # hash of card params. params may also be provided individually or added interactively.
          [--number | --card-number=CARD_NUMBER]          # credit card number. usually 16 digits long
          [--exp-month | --card-exp-month=CARD_EXP_MONTH] # Two digit expiration month of card
          [--exp-year | --card-exp-year=CARD_EXP_YEAR]    # Four digit expiration year of card
          [--cvc | --card-cvc=CARD_CVC]                   # Three or four digit security code located on the back of card
          [--name | --card-name=CARD_NAME]                # Cardholder's full name as displayed on card
          [--account | --bank-account=key:value]          # hash of account params. params may also be provided individually or added interactively.
          [--country=COUNTRY]                             # country of bank account. currently supports 'US' only
          [--routing-number=ROUTING_NUMBER]               # ACH routing number for bank account
          [--account-number=ACCOUNT_NUMBER]               #

    create a new token of type TYPE(card or account)

#### Token Find

    Usage:
      stripe token find ID

    Find a Token
### Cards

    Commands:
      stripe cards create --owner=OWNER     # Create a new card for OWNER (customer or recipient)
      stripe cards delete ID --owner=OWNER  # Delete ID card for OWNER (customer or recipient)
      stripe cards find ID --owner=OWNER    # Find ID card for OWNER (customer or recipient)
      stripe cards help [COMMAND]           # Describe subcommands or one specific subcommand
      stripe cards list --owner=OWNER       # List cards for OWNER (customer or recipient)

#### Card Create

    Usage:
      stripe card create --customer, --recipient, --owner=OWNER

    Options:
          [--token | --card=CARD]                     # credit card Token or ID. May also be created interactively.
          [--number | --card-number=CARD_NUMBER]
          [--exp-month | --card-exp-month=EXP_MONTH]  # Two digit expiration month of card
          [--exp-year | --card-exp-year=EXP_YEAR]     # Four digit expiration year of card
          [--cvc | --card-cvc=CARD_CVC]               # Three or four digit security code located on the back of card
          [--name | --card-name=CARD_NAME]            # Cardholder's full name as displayed on card
          --customer, --recipient, --owner=OWNER      # id of customer or recipient receiving new card

    create a new card for OWNER (customer or recipient)

#### Card Delete

    Usage:
      stripe card delete ID --owner=OWNER

    Options:
          --owner=OWNER                      # id of customer or recipient to search within

    delete ID card for OWNER (customer or recipient)

#### Card Find

    Usage:
      stripe card find ID --owner=OWNER

    Options:
          --owner=OWNER                      # id of customer or recipient to search within

    find ID card for OWNER (customer or recipient)

#### Cards List

    Usage:
      stripe cards list --owner=OWNER

    Options:
          [--starting-after=STARTING_AFTER]  # The ID of the last object in the previous paged result set. For cursor-based pagination.
          [--ending-before=ENDING_BEFORE]    # The ID of the first object in the previous paged result set, when paging backwards through the list.
          [--limit=LIMIT]                    # a limit on the number of resources returned, between 1 and 100
          [--offset=OFFSET]                  # the starting index to be used, relative to the entire list
          [--count=COUNT]                    # deprecated: use limit
          --owner=OWNER                      # id of customer or recipient to search within

    List cards for OWNER (customer or recipient)

### Customers

    Commands:
      stripe customers create          # Create a new customer
      stripe customers delete ID       # Delete a customer
      stripe customers find ID         # Find a customer
      stripe customers help [COMMAND]  # Describe subcommands or one specific subcommand
      stripe customers list            # List customers

#### Customer Create

    Usage:
      stripe customer create

    Options:
          [--description=DESCRIPTION]                     # Arbitrary description to be displayed in Stripe Dashboard.
          [--email=EMAIL]                                 # Customer's email address. Will be displayed in Stripe Dashboard.
          [--plan=PLAN]                                   # The ID of a Plan this customer should be subscribed to. Requires a credit card.
          [--coupon=COUPON]                               # The ID of a Coupon to be applied to all of Customer's recurring charges
          [--quantity=QUANTITY]                           # A multiplier for the plan option. defaults to `1'
          [--trial-end=TRIAL_END]                         # apply a trial period until this date. Override plan's trial period.
          [--account-balance=ACCOUNT_BALANCE]             # customer's starting account balance in cents. A positive amount will be added to the next invoice while a negitive amount will act as a credit.
          [--token | --card=CARD]                         # credit card Token or ID. May also be created interactively.
          [--number | --card-number=CARD_NUMBER]          # credit card number. usually 16 digits long
          [--exp-month | --card-exp-month=CARD_EXP_MONTH] # Two digit expiration month of card
          [--exp-year | --card-exp-year=CARD_EXP_YEAR]    # Four digit expiration year of card
          [--cvc | --card-cvc=CARD_CVC]                   # Three or four digit security code located on the back of card
          [--name | --card-name=CARD_NAME]                # Cardholder's full name as displayed on card
          [--metadata=key:value]                          # a key/value store of additional user-defined data

    Create a new customer

#### Customer Delete

    Usage:
      stripe customer delete ID

    Delete a customer

#### Customer Find

    Usage:
      stripe customer find ID

    Find a customer

#### Customers List

    Usage:
      stripe customers list

    Options:
          [--starting-after=STARTING_AFTER]  # The ID of the last object in the previous paged result set. For cursor-based pagination.
          [--ending-before=ENDING_BEFORE]    # The ID of the first object in the previous paged result set, when paging backwards through the list.
          [--limit=LIMIT]                    # a limit on the number of resources returned, between 1 and 100
          [--offset=OFFSET]                  # the starting index to be used, relative to the entire list
          [--count=COUNT]                    # deprecated: use limit

    List customers


### Subscriptions

    Commands:
      stripe subscriptions cancel ID --customer=CUSTOMER      # cancel ID subscription for CUSTOMER customer
      stripe subscriptions create --customer=CUSTOMER         # create a subscription for CUSTOMER customer
      stripe subscriptions find ID --customer=CUSTOMER        # find ID subscription for CUSTOMER customer
      stripe subscriptions help [COMMAND]                     # Describe subcommands or one specific subcommand
      stripe subscriptions list --customer=CUSTOMER           # List subscriptions for CUSTOMER customer
      stripe subscriptions reactivate ID --customer=CUSTOMER  # reactivate auto-renewal if `cancel-at-period-end` was set to true

#### Subscription Cancel

    Usage:
      stripe subscription cancel ID c, --customer=CUSTOMER

    Options:
          [--at-period-end], [--no-at-period-end]         # delay cancellation until end of current period, default: false
          c, --customer=CUSTOMER                          # ID of customer who's subscription we want to cancel

    cancel ID subscription for CUSTOMER customer

#### Subscription Create

    Usage:
      stripe subscription create c, --customer=CUSTOMER

    Options:
          [--plan=PLAN]                                   # the plan to assign to CUSTOMER customer
          [--coupon=COUPON]                               # id of a coupon to apply
          [--trial-end=TRIAL_END]                         # apply a trial period until this date. Override plan's trial period.
          [--token | --card=CARD]                         # credit card Token or ID. May also be created interactively.
          [--number | --card-number=CARD_NUMBER]          # credit card number. usually 16 digits long
          [--exp-month | --card-exp-month=CARD_EXP_MONTH] # Two digit expiration month of card
          [--exp-year | --card-exp-year=CARD_EXP_YEAR]    # Four digit expiration year of card
          [--cvc | --card-cvc=CARD_CVC]                   # Three or four digit security code located on the back of card
          [--name | --card-name=CARD_NAME]                # Cardholder's full name as displayed on card
          [--metadata=key:value]                          # a key/value store of additional user-defined data
          -c, --customer=CUSTOMER                         # ID of customer receiving the new subscription

    create a subscription for CUSTOMER customer

#### Subscription Find

    Usage:
      stripe subscription find ID -c, --customer=CUSTOMER

    Options:
          -c, --customer=CUSTOMER             # ID of customer to search within

    find ID subscription for CUSTOMER customer

#### Subscriptions List

    Usage:
      stripe subscriptions list -c, --customer=CUSTOMER

    Options:
          [--starting-after=STARTING_AFTER]   # The ID of the last object in the previous paged result set. For cursor-based pagination.
          [--ending-before=ENDING_BEFORE]     # The ID of the first object in the previous paged result set, when paging backwards through the list.
          [--limit=LIMIT]                     # a limit on the number of resources returned, between 1 and 100
          [--offset=OFFSET]                   # the starting index to be used, relative to the entire list
          [--count=COUNT]                     # deprecated: use limit
          -c, --customer=CUSTOMER             # ID of customer who's subscriptions we want to list

    List subscriptions for CUSTOMER customer

#### Subscription Reactivate

    Usage:
      stripe subscription reactivate ID -c, --customer=CUSTOMER

    Options:
          -c, --customer=CUSTOMER         # ID of customer who's subscription we want to reactivate

    reactivate auto-renewal if `cancel-at-period-end` was set to true

### Invoices

    Commands:
      stripe invoices close ID            # Close an unpaid invoice
      stripe invoices find ID             # Find an invoice
      stripe invoices help [COMMAND]      # Describe subcommands or one specific subcommand
      stripe invoices list                # List invoices (optionally by customer_id)
      stripe invoices pay ID              # trigger an open invoice to be paid immediately
      stripe invoices upcoming CUSTOMER   # find the upcoming invoice for CUSTOMER

#### Invoice Close

    Usage:
      stripe invoice close ID

    close an unpaid invoice

#### Invoice Find

    Usage:
      stripe invoice find ID

    Find an invoice

#### Invoices List

    Usage:
      stripe invoice list

    Options:
          [--starting-after=STARTING_AFTER]  # The ID of the last object in the previous paged result set. For cursor-based pagination.
          [--ending-before=ENDING_BEFORE]    # The ID of the first object in the previous paged result set, when paging backwards through the list.
          [--limit=LIMIT]                    # a limit on the number of resources returned, between 1 and 100
          [--offset=OFFSET]                  # the starting index to be used, relative to the entire list
          [--count=COUNT]                    # deprecated: use limit
          [--customer=CUSTOMER]              # a customer ID to filter results by

    List invoices (optionally by customer_id)

#### Invoice Pay

    Usage:
      stripe invoice pay ID

    trigger an open invoice to be paid immediately

#### Invoice Upcoming

    Usage:
      stripe invoice upcoming CUSTOMER

    find the upcoming invoice for CUSTOMER

### Invoice Items

    Commands:
      stripe invoice-items create          # Create a new invoice item for customer
      stripe invoice-items delete ID       # Delete a invoice item
      stripe invoice-items find ID         # Find an invoice item
      stripe invoice-items help [COMMAND]  # Describe subcommands or one specific subcommand
      stripe invoice-items list            # List invoice items (optionally by customer_id)


#### Invoice Item Create

    Usage:
      stripe invoice-item create

    Options:
          [--customer=CUSTOMER]                      # The ID of customer for the invoice item
          [--amount=N]                               # Amount in dollars (or cents when --no-dollar-amounts)
          [--currency=CURRENCY]                      # 3-letter ISO code for currency
                                                     # Default: usd
          [--description=DESCRIPTION]                # Arbitrary description of charge
          [--invoice=INVOICE]                        # The ID of an existing invoice to add this invoice item to
          [--subscription=SUBSCRIPTION]              # The ID of a subscription to add this invoice item to
          [--discountable], [--no-discountable]      # Controls whether discounts apply to this invoice item
          [--metadata=key:value]                     # A key/value store of additional user-defined data

    Create a new invoice item for customer

#### Invoice Item Delete

    Usage:
      stripe invoice-item delete ID

    Delete a invoice item

#### Invoice Item Find

    Usage:
      stripe invoice-item find ID

    Find an invoice item

#### Invoice Item List

    Usage:
      stripe invoice-items list

    Options:
          [--starting-after=STARTING_AFTER]          # The ID of the last object in the previous paged result set. For cursor-based pagination.
          [--ending-before=ENDING_BEFORE]            # The ID of the first object in the previous paged result set, when paging backwards through the list.
          [--limit=LIMIT]                            # a limit on the number of resources returned, between 1 and 100
          [--offset=OFFSET]                          # the starting index to be used, relative to the entire list
          [--count=COUNT]                            # deprecated: use limit
          [--customer=CUSTOMER]                      # ID of customer who's invoice items we want to list

    List invoice items (optionally by customer_id)

### Plans

    Commands:
      stripe plans create          # Create a new plan
      stripe plans delete ID       # Delete a plan
      stripe plans find ID         # Find a plan
      stripe plans help [COMMAND]  # Describe subcommands or one specific subcommand
      stripe plans list            # List plans

#### Plan Create

    Usage:
      stripe plans create

    Options:
          [--name=NAME]            # Name displayed on invoices and in the Dashboard
          [--id=ID]                # Unique name used to identify this plan when subscribing a customer

changelog.md
# Stripe-CLI Project Change Log

## 1.8.3

 - new command-line option `--expand` provides support for Stripe API expansions
   - https://stripe.com/docs/api#expanding_objects
   - stripe-ruby dependancy updated to version 1.24.0

## 1.8.2

 - bug fixes
   - it was previously possible for certain command-line arguments to be passed through to the Stripe API resulting in API exceptions.
   - creating tokens would fail for bank account type

## 1.8.1

 - new configuration option to strip nil-valued attributes from command-line output
   - use on command-line with global flag `--strip-nils`
   - or set in config-file with `strip_nils=true`
   - override config-file setting with global command-line flag `--no-strip-nils`

## 1.8.0

 - top-level `invoice_items` command
   - w/ actions: `find`, `list`, `create`, & `delete`

## 1.7.0

- new `dates` config option for specifying how dates should be displayed (`utc`, `unix`, or `local`)

## 1.6.4

 - better descriptions for ALL operations
 - Bug fix
 - update `transfers` command options
   - `--statement-description`
   - `--bank-account`
   - `--card`

## 1.6.3

 - new `dollar_amounts` config setting for toggling expected currency units between **dollars** and **cents**
 - `--no-dollar-amounts` global option flag
 - `--dollar-amounts` global option flag

## 1.6.1

 - extensive option alias support throughout, with more on the way
 - more extensive documentation in README, including subcommand descriptions. (still incomplete)
 - add TOC in README

## 1.6.0

 - token creation correctly handles credit card params in any combination of individual options and --card=key:value pairs and still collects any missing params via interactive menu.
 - any operation that accepts a credit card ID or token also accepts credit card parameters directly and incorporates the credit card interactive menu.
 - improved option descriptions in all operations of all commands

## 1.5.2

 - tack **'/Stripe-CLI vX.X.X'** onto user-agent header of every api request
   - **Disclaimer:** though I believe it to be completely harmless in this instance, this feature involves monkey-patching the `Stripe` module with an `alias_method_chain` style wrapper around `Stripe#request_headers`
   - The Stripe Dashboard depends on the user-agent to distinguish the **source** of each api request
   - this helps differentiate between requests made by application code and those made using the Stripe-CLI gem


## 1.5.1

 - fix unreported bug where Stripe::ListObjects were not displayed as simple arrays by awesome_print
   - only present in conjunction with stripe-ruby version 1.14 and up
   - due to a change in stripe-ruby by which `Stripe::StripeObject#to_hash` now calls itself recursivly
   - resort to using `Object#instance_variable_get` to peek at a StripeObject's `@values` variable which already contains a hash of its attributes
     - I am open to suggestions for a less brittle solution

## 1.5.0

 - Refunds as a top-level command

## 1.4.6

 - subscriptions `reactivate` subcommand
 - `--at-period-end` flag added to `subscriptions cancel`
 - `--refund-application-fee` flag added to `charges refund`
 - add `--metadata` to the remaining subcommands that need it

## 1.4.4

 - `--receipt_email` and `--statement_description` options added to `charges create`

## 1.4.3

 - top-level `version` command for printing current gem version
   - hidden from standard `help` banner
   - aliased as `-v` and `--version`

## 1.4.1

 - per project config files (.stripecli)

## 1.4.0

 - credit cards as a top-level command
   - for customers and recipients
   - w/ actions: find, list, create, & delete

## 1.3.0

  - introduce multiple subscriptions per customer
    - w/ actions: cancel, create, find, & list

## 1.2.2

  - introduce optional cursor-based pagination options to all `list` operations

lib/stripe/cli/commands/charges.rb
module Stripe
  module CLI
    module Commands
      class Charges < Command
        include Stripe::Utils

        desc "list", "List charges (optionally by customer_id)"
        option :starting_after, :desc => "The ID of the last object in the previous paged result set. For cursor-based pagination."
        option :ending_before, :desc => "The ID of the first object in the previous paged result set, when paging backwards through the list."
        option :limit, :desc => "a limit on the number of resources returned, between 1 and 100"
        option :offset, :desc => "the starting index to be used, relative to the entire list"
        option :count, :desc => "deprecated: use limit"
        option :customer, :desc => "ID of customer who's charges we want to list"
        def list
          super Stripe::Charge, options
        end

        desc "find ID", "Find a charge"
        def find charge_id
          super Stripe::Charge, charge_id
        end

        desc "refund ID", "Refund a charge"
        option :amount, :type => :numeric, :desc => "Refund amount in dollars. (or cents when --no-dollar-amounts) defaults to entire charge"
        option :metadata, :type => :hash, :desc => "a key/value store of additional user-defined data"
        option :refund_application_fee, :type => :boolean, :default => false, :desc => "Whether or not to refund the application fee"
        def refund charge_id
          options[:amount] = convert_amount(options[:amount]) if options[:amount]
          if charge = retrieve_charge(charge_id)
            request charge.refunds, :create, options
          end
        end

        desc "capture ID", "Capture a charge"
        def capture charge_id
          request Stripe::Charge.new(charge_id, api_key), :capture
        end

        desc "create", "Create a charge"
        option :customer, :desc => "The ID of an existing customer to charge"
        option :card, :aliases => "--token", :desc => "credit card Token or ID. May also be created interactively."
        option :card_number, :aliases => "--number", :desc => "credit card number. usually 16 digits long"
        option :card_exp_month, :aliases => "--exp-month", :desc => "Two digit expiration month of card"
        option :card_exp_year, :aliases => "--exp-year", :desc => "Four digit expiration year of card"
        option :card_cvc, :aliases => "--cvc", :desc => "Three or four digit security code located on the back of card"
        option :card_name, :aliases => "--name", :desc => "Cardholder's full name as displayed on card"
        option :amount, :type => :numeric, :desc => "Charge amount in dollars (or cents when --no-dollar-amounts)"
        option :currency, :default => "usd", :desc => "3-letter ISO code for currency"
        option :description, :desc => "Arbitrary description of charge"
        option :capture, :type => :boolean, :default => true, :desc => "Whether or not to immediately capture the charge. Uncaptured charges expire in 7 days"
        option :metadata, :type => :hash, :desc => "A key/value store of additional user-defined data"
        option :statement_description, :desc => "Displayed alongside your company name on your customer's card statement (15 character max)"
        option :receipt_email, :desc => "Email address to send receipt to. Overrides default email settings."
        def create
          options[:amount] = convert_amount(options[:amount])

          options[:card] ||= credit_card( options ) unless options[:customer]

          super Stripe::Charge, options
        end

      end
    end
  end
end